thread.c

```diff
@@ -1108,129 +1122,152 @@ rb_thread_create_ractor(rb_ractor_t *g, VALUE args, VALUE proc)
 
 
 struct join_arg {
-    rb_thread_t *target, *waiting;
-    rb_hrtime_t *limit;
+    struct rb_waiting_list *waiting_list;
+    rb_thread_t *target;
+    VALUE timeout;
 };
 
 static VALUE
 remove_from_join_list(VALUE arg)
 {
     struct join_arg *p = (struct join_arg *)arg;
-    rb_thread_t *target_th = p->target, *th = p->waiting;
+    rb_thread_t *target_thread = p->target;
 
-    if (target_th->status != THREAD_KILLED) {
-	rb_thread_list_t **p = &target_th->join_list;
+    if (target_thread->status != THREAD_KILLED) {
+        struct rb_waiting_list **join_list = &target_thread->join_list;
 
-	while (*p) {
-	    if ((*p)->th == th) {
-		*p = (*p)->next;
-		break;
-	    }
-	    p = &(*p)->next;
-	}
+        while (*join_list) {
+            if (*join_list == p->waiting_list) {
```

Comment:
`p->waiting_list` seems confusing here. We just want to compare against a thread.

Reply:
No, we only want to remove the entry for the given linked list node, which has a thread and fiber pair. IF two fibers were waitin on the same thread, that would be two different entries in the list.